tests/test_utils.py

```diff
@@ -228,3 +235,101 @@ def test_confirm_tags_reject_raises_abort(confirm):
+
+
+# build_csv
+
+def test_build_csv_empty_data():
+    assert build_csv([]) == ''
+
+
+def test_build_csv_one_col():
+    lt = csv.get_dialect('excel').lineterminator
+    data = [{'col': 'value'}, {'col': 'another value'}]
+    result = lt.join(['col', 'value', 'another value']) + lt
+    assert build_csv(data) == result
+
+
+def test_build_csv_multiple_cols():
+    lt = csv.get_dialect('excel').lineterminator
+    dm = csv.get_dialect('excel').delimiter
+    data = [
+        co.OrderedDict([('col1', 'value'),
+                        ('col2', 'another value'),
+                        ('col3', 'more')]),
+        co.OrderedDict([('col1', 'one value'),
+                        ('col2', 'two value'),
+                        ('col3', 'three')])
+    ]
+    result = lt.join([
+        dm.join(['col1', 'col2', 'col3']),
+        dm.join(['value', 'another value', 'more']),
+        dm.join(['one value', 'two value', 'three'])
+        ]) + lt
+    assert build_csv(data) == result
+
+
+# frames_to_csv
+
+def test_frames_to_csv_empty_data(watson):
+    assert frames_to_csv(watson.frames) == ''
+
+
+def test_frames_to_csv(watson):
+    watson.start('foo', tags=['A', 'B'])
+    watson.stop()
+
+    result = frames_to_csv(watson.frames)
+
+    read_csv = list(csv.reader(StringIO(result)))
+    header = ['id', 'start', 'stop', 'project', 'tags']
+    assert len(read_csv) == 2
+    assert read_csv[0] == header
+    assert read_csv[1][3] == 'foo'
+    assert read_csv[1][4] == 'A, B'
+
+
+# frames_to_json
+
+def test_frames_to_json_empty_data(watson):
+    assert frames_to_json(watson.frames) == '[]'
+
+
+def test_frames_to_json(watson):
+    watson.start('foo', tags=['A', 'B'])
+    watson.stop()
+
+    result = json.loads(frames_to_json(watson.frames))
+
+    keys = {'id', 'start', 'stop', 'project', 'tags'}
+    assert len(result) == 1
+    assert set(result[0].keys()) == keys
+    assert result[0]['project'] == 'foo'
+    assert result[0]['tags'] == ['A', 'B']
+
+
+# flatten_report_for_csv
+
+def test_flatten_report_for_csv(watson):
+    now = arrow.utcnow().ceil('hour')
+    watson.add('foo', now.shift(hours=-4), now, ['A', 'B'])
+    watson.add('foo', now.shift(hours=-5), now.shift(hours=-4), ['A'])
+    watson.add('foo', now.shift(hours=-7), now.shift(hours=-5), ['B'])
+
+    start = now.shift(days=-1)
+    stop = now
+    result = flatten_report_for_csv(watson.report(start, stop))
+
+    assert len(result) == 3
+    for row in result:
+        assert row['from'] == start.format('YYYY-MM-DD 00:00:00')
+        assert row['to'] == stop.format('YYYY-MM-DD 23:59:59')
+        assert row['project'] == 'foo'
+        if row['tag'] == 'A':
+            assert row['time'] == (4 + 1) * 3600
+        elif row['tag'] == 'B':
+            assert row['time'] == (4 + 2) * 3600
+        elif row['tag'] == '':
+            assert row['time'] == (4 + 1 + 2) * 3600
+        else:
+            assert "Invalid tag found" and False
```

Comment:
Although it would be a bit more verbose, I think this tests should be written without the loop. I'm never 100% confident with loops & conditions in the tests :) .